File: uniffi_bindgen/src/bindings/kotlin/templates/HandleMap.kt
// Initial value and increment amount for handles. 
// These ensure that Kotlin-generated handles always have the lowest bit set
private const val UNIFFI_HANDLEMAP_INITIAL = 1.toLong()
private const val UNIFFI_HANDLEMAP_DELTA = 2.toLong()

// Map handles to objects
//
// This is used pass an opaque 64-bit handle representing a foreign object to the Rust code.
internal class UniffiHandleMap<T: Any> {
    private val map = ConcurrentHashMap<Long, T>()
    // Start 
    private val counter = java.util.concurrent.atomic.AtomicLong(UNIFFI_HANDLEMAP_INITIAL)

    val size: Int
        get() = map.size

    // Insert a new object into the handle map and get a handle for it
    fun insert(obj: T): Long {
        val handle = counter.getAndAdd(UNIFFI_HANDLEMAP_DELTA)
        map.put(handle, obj)
        return handle
    }

    // Clone a handle, creating a new one
    fun clone(handle: Long): Long {
        val obj = map.get(handle) ?: throw InternalException("UniffiHandleMap.clone: Invalid handle")
        return insert(obj)
    }

    // Get an object from the handle map
    fun get(handle: Long): T {
        return map.get(handle) ?: throw InternalException("UniffiHandleMap.get: Invalid handle")
    }

    // Remove an entry from the handlemap and get the Kotlin object back
    fun remove(handle: Long): T {
        return map.remove(handle) ?: throw InternalException("UniffiHandleMap: Invalid handle")
    }
}

File: uniffi_bindgen/src/bindings/ruby/templates/HandleMap.rb
# A handle map for converting Ruby objects to/from integer handles.
#
# When passing callback interface or trait interface objects to Ruby, we store
# the Ruby object in this map and pass the integer handle to Rust. When Rust
# calls back, we look up the Ruby object by its handle.
#
# Ruby-created handles are always odd (staring at 1, incrementing by 2).
# Rust-created handles are always even. This convention lets us distinguish
# the source in lift/lower without extra metadata.


class UniffiHandleMap
  def initialize
    @lock = Monitor.new
    @map = {}
    @counter = 1 # Start at 1 (odd), increment by 2
  end

  def insert(obj)
    @lock.synchronize do
      handle = @counter
      @counter += 2
      @map[handle] = obj
      handle
    end
  end

  def get(handle)
    @lock.synchronize do
      obj = @map[handle]
      raise InternalError, "UniffiHandleMap.get: invalid handle #{handle}" if obj.nil?
      obj
    end
  end

  def clone_handle(handle)
    @lock.synchronize do
      obj = @map[handle]
      raise InternalError, "UniffiHandleMap.clone_handle: invalid handle #{handle}" if obj.nil?
      new_handle = @counter
      @counter += 2
      @map[new_handle] = obj
      new_handle
    end
  end

  def remove(handle)
    @lock.synchronize do
      obj = @map.delete(handle)
      raise InternalError, "UniffiHandleMap.remove: invalid handle #{handle}" if obj.nil?
      obj
    end
  end

  # Used for testing/debugging purposes
  def size
    @lock.synchronize { @map.size }
  end
end

private_constant :UniffiHandleMap
